### DISCUSSION TITLE
Formatting our Python code

### DISCUSSION THREAD

**Discussion by ckreibich** [zeek/zeek#2602] [Projects]
We have a bunch of Python code scattered across our projects, with some obvious heavy-weights like `zkg`, `zeekctl`, and `zeekscript`. These do not currently enforce code formatting. It makes sense to standardize formatting for them going forward, and I wanted to discuss this real quick so we're all on the same page, can reference this in the future, and don't end up sneaking these in differently repo-by-repo.

Seems there are two contenders, [yapf](https://github.com/google/yapf) and [black](https://github.com/psf/black). The former allows customization, similarly to `clang-format`, the latter has one style.

We currently have some use of yapf, as follows:
- `zeek/zeek`  and `zeek/binpac` have a basic `.style.yapf` and pre-commit setup
- `zeek/bifcl` has a pre-commit setup
- `zeek/btest`  has a slightly different `.style.yapf` and pre-commit setup

Shout if I've overlooked anything here. (The `out_ptr` submodule has customized use of `yapf`, but it points straight upstream.)

The other day @bbannier proposed use of `black` in the `zkg` package template, so I'm wondering:
- Is there a reason why we'd want to use multiple formatters?
- If one, which? I don't feel strongly about this, but I'm wondering if other folks do.

I'd probably leave `zeekctl` alone at this point, fwiw. Also, it seems we can remove the pre-commit setups from bifcl and binpac? I don't immediately see any Python use in them.

Thoughts welcome...

**Comment by bbannier** [comment]
For context, I the [PR originally proposing Black for zeek/package-template](https://github.com/zeek/package-template/pull/6#discussion_r1022060737) my preference for Black came from it being very aggressive:

> I introduced this to remove the formatting element from reviews completely (at least for Python files). I don't care much about the particular style, but went with Black since it is most aggressive in formatting; pure formatting changes to a Black-formatted file likely would result in the edits being removed again while e.g., yapf might leave them in.

I perceive that it has no configuration options as another benefit (I understand that some might perceive this as a negative); it now has a stable style for which they make [stability guarantees](https://black.readthedocs.io/en/stable/the_black_code_style/index.html#stability-policy). In the past I got the impression that the style enforced by Black causes smaller diffs when e.g., adding additional keyword arguments, or adding list fields.

That being said, I believe that in world where code is formatted automatically the particular style becomes less interesting so different subprojects should be able use their own style without harm.

> Also, it seems we can remove the pre-commit setups from bifcl and binpac? I don't immediately see any Python use in them.

There is an [empty `setup.py` in binpac](https://github.com/zeek/binpac/blob/d5433c8071a3bab58ab08ff5a141d12d106711db/setup.py) which might have ended there by accident; bifcl has no Python files. FYI, if no Python files are present `pre-commit` would download the formatter code, but always skip its execution though since the formatter executes only on Python files (which wouldn't be there).

**Comment by awelzel** [comment]
I've used `black` before. I've not used `yapf`. IMO no configuration is a huge plus for `black`.

Not sure that's adding much, but y personal pet peeve is string quoting style: `'` vs `"` and I get hung-up on it easily. `yapf` doesn't seem to be able to enforce a style here (https://github.com/google/yapf/issues/399). Based on that extremely narrow/limited view I'd select `black`.

Probably should consider adding `isort` for sorting imports.

**Comment by ckreibich** [comment]
I've toyed a bit with both and I'm liking `black` too, both in terms of the results and lack of (fanciful) options. Here's a preview of what a reformatted `zkg` will look like — it's mostly trivial: https://github.com/ckreibich/package-manager/pull/5/files

Is there any reason why we wouldn't swap yapf for black in the places where we currently have the former?

> Probably should consider adding isort for sorting imports.

Yep, I like that too.

**Comment by bbannier** [comment]
> Is there any reason why we wouldn't swap yapf for black in the places where we currently have the former?

When I originally added the yapf setup there was no consensus on the preferred style, so I tried to produce minimal diffs; the only yapf setting we override for zeek/zeek is going with an increased line length (copied from zeek/btest). If we committed to a style and a tool like Black to consistently enforce it, doing more extensive reformatting once is more than likely worth it.

**Comment by ckreibich** [comment]
Cool. I am all for using black everywhere if nobody voices strong objections.
